### PR TITLE
[3.8] Add SodiumCipher to support libsodium (no polyfill option)

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes

This is an alternative to https://github.com/joomla/joomla-cms/pull/16754 without support for the polyfill library.

PHP 7.2 will be removing `ext/mcrypt` from the core distribution and `ext/sodium` will be added to core as a new encryption library.  As the `JCryptCipher` classes are primarily built around `mcrypt` (and inherently what is present in PHP core), we should add support for the new library as well.

`ext/sodium` isn't restricted to PHP 7.2 installations only.  There is also a [PECL extension](https://pecl.php.net/package/libsodium) providing support for PHP 5.4 through 7.1 and the [sodium_compat polyfill](https://github.com/paragonie/sodium_compat) providing support down to PHP 5.2.4.  The polyfill is not included with this pull request due to concerns expressed with using security related code that had not received a proper audit.

### Testing Instructions

The unit test coverage for the class best demonstrates its use:

```php
$cipher = new \Joomla\CMS\Crypt\Cipher\SodiumCipher;
$key    = $cipher->generateKey();
$data   = 'My encrypted data.';

if (version_compare(PHP_VERSION, '7.2', '>='))
{
	$cipher->setNonce(sodium_randombytes_buf(SODIUM_CRYPTO_BOX_NONCEBYTES));
}
else
{
	$cipher->setNonce(\Sodium\randombytes_buf(\Sodium\CRYPTO_BOX_NONCEBYTES));
}

$encrypted = $cipher->encrypt($data, $key);
$decrypted = $cipher->decrypt($encrypted, $key);

if ($decrypted !== $data)
{
	throw new RuntimeException('The data was not decrypted correctly.');
}
```

One thing to note here.  Unlike our other ciphers, a [nonce](https://en.wikipedia.org/wiki/Cryptographic_nonce) must be set and used for both the encryption and decryption of data, and must be the same value on both ends of the process.  Otherwise, this is no different than the existing API.

### Documentation Changes Required

Document the added class and its use.  Note that use is restricted to PHP 5.4+ and requires the `libsodium` extension (either from PECL on PHP 5.4 through 7.1 or core PHP on 7.2).

//cc @joomla/security

### CODE REVIEW

**Comment by SniperSister** [comment]
Makes sense from my side, +1 from the JSST

File: libraries/vendor/composer/autoload_classmap.php [hunk 1]
```diff
@@ -8,12 +8,208 @@
 return array(
     'CallbackFilterIterator' => $vendorDir . '/joomla/compat/src/CallbackFilterIterator.php',
     'EasyPeasyICS' => $vendorDir . '/phpmailer/phpmailer/extras/EasyPeasyICS.php',
+    'Joomla\\Application\\AbstractApplication' => $vendorDir . '/joomla/application/src/AbstractApplication.php',
```
Comment (laoneo) [inline libraries/vendor/composer/autoload_classmap.php]: Why did the composer files change?

Comment (mbabker) [reply libraries/vendor/composer/autoload_classmap.php]: I started from the other PR and rolled back adding the Composer package, I also ran with the `-o` option.

Comment (laoneo) [reply libraries/vendor/composer/autoload_classmap.php]: I would rather do the optimisation after #16896 get merged from staging into 3.8 and not in a pr which has another purpose.

**Comment by zero-24** [comment]
I think this can be closed based on the feedback in the other PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16917">issues.joomla.org/tracker/joomla-cms/16917</a>.</sub>

**Comment by wilsonge** [comment]
Indeed it can be

**Pull request closed without merging**